### PR TITLE
WIP: Store file contents alongside VersionEntries

### LINKED ISSUE #248: VersionFile should include file type
As discussed at https://github.com/mozilla/addons-code-manager/pull/243#discussion_r259045468, we probably want a `VersionFile` to know what file type it is, so that we know whether we need to display a text file, an image, or whether it is undisplayable. 

I originally thought we could use `VersionEntryType`, but that is calculated based on the data we have in the file tree, and not on the data returned by the API for the file itself, so we probably need another way to calculate the type of a file.

Looking further, it seems that we don't get any of the information we need in the API response for the file itself, other than the file name/path which includes an extension so we could possibly use that.

The file tree does contain "file type" information for each of the entries, which we are currently using to calculate `VersionEntryType`, and it also includes a `mimetype` which might also be useful for this use case.

So, looking at all of that, it seems like the app should make assumptions about the type of a file based on the data in the file tree (which is part of `Version`) and not based on the data of the file itself (which is part of `VersionFile`). Given that, I wonder if it would make sense to change the shape of the reducer so that all data about a file, including its content, is stored in `VersionEntry`, which is stored in the `entries` property of a `Version`. This would mean the reducer state would no longer have properties for both `versionInfo` and `versionFiles`, and would only have a property for `versions`, which would hold `Version`s, which would in turn hold `VersionEntry`s inside its `entries` property.

Am I making sense? I can put forward a POC if it would make it clearer.

@willdurand / @kumar303 what do you think?

### PR DESCRIPTION
**Pull request by bobsilverberg**
Fixes #248 

This ended up being a pretty big patch, but the work is mostly done now, so if we like this approach I think we can move forward with it. I do like the new shape of the reducer, and I think it's a bit friendlier to work with from a consumer standpoint, but the tests have become a bit more complicated.

I can do some work to clean the tests up a bit, perhaps with some helpers if we do decide to go this route.

For now perhaps it's best to just focus on the changes to the reducer and the consumers of the reducer, and if everyone is happy with this approach then we can look at the tests as well.

### CODE REVIEW

File: src/reducers/versions.tsx [hunk 1]
```diff
@@ -89,6 +91,7 @@ export type VersionFile = {
 
 type VersionEntry = {
   depth: number;
+  file: VersionFile | null;
```
Comment (kumar303) [inline src/reducers/versions.tsx]: We've had a lot of trouble in the past dealing with fields like this that may or may not be populated with an object. This is an indication to me that `file` does not belong here. Can we use another mapping?

**Comment by bobsilverberg** [comment]
Well, I did spend a lot of time on this, but it seems like we have higher priorities, and @willdurand's workaround seems fine for now, so I guess we can just shelve it and perhaps revisit it in the future. I'm fine with that.

**Comment by bobsilverberg** [comment]
I'm going to take a look at just adding some code to the reducer so that we have access to the entry data whenever we request a file, which is what @willdurand's workaround does. Adding this code into the reducer, instead of into the consumers seems like a good idea as we may need it in more than one place.

That should drastically reduce the size of the patch and shouldn't affect tests, etc. as this current patch does.

**Comment by kumar303** [comment]
> I'm going to take a look at just adding some code to the reducer so that we have access to the entry data whenever we request a file

Thanks. I think that will be a great place to start. We may come back to this patch here later.

**Comment by bobsilverberg** [comment]
I'll close this PR and try out the smaller idea.

**Pull request closed without merging**